### DISCUSSION TITLE
Visualizing an array calculated with paraview in Trame

### DISCUSSION THREAD

**Discussion by smkacur** [Kitware/trame#338] [Trame/ParaView] [answered]
Hello,

I have data in the form of a VTK timeseries and would like to calculate the difference from t0 for one of the arrays in the VTK using Paraview. However, when I try to visualize the newly calculated array: ‘d_accl’, I receive this error: 

`[paraview]vtkSMPVRepresentationPr:288   WARN| vtkSMPVRepresentationProxy (0x166508530): Could not determine array range.
`

When I load the vtk timeseries into trame I am able to visualize original arrays and change which array the vtks are colored by. I have also confirmed that the function to calculate the difference from t0 works in Paraview, so I am wondering if there is a step I am missing where the newly calculated array needs to be added back to the original VTK data? 

In order to make things easier I have attempted to recreate the issue using the [Paraview/TimeAnimation](https://github.com/Kitware/trame/blob/master/examples/07_paraview/TimeAnimation/app.py) example. The current example throws the same error as above when coloring by any of the arrays, I guess there is an issue reading in the arrays from the data/can.json. For reference, here is the amended script that I have been using, which contains the function to calculate the difference from t0. 

```python
import paraview.web.venv  # Available in PV 5.10-RC2+
import os
import json
import asyncio

from trame.app import get_server, asynchronous
from trame.widgets import vuetify, paraview
from trame.ui.vuetify import SinglePageLayout

import paraview as pv
from paraview import simple

# -----------------------------------------------------------------------------
# Trame setup
# -----------------------------------------------------------------------------

server = get_server()
state, ctrl = server.state, server.controller

# -----------------------------------------------------------------------------

animation_scene = simple.GetAnimationScene()
time_keeper = animation_scene.TimeKeeper

metadata = None
time_values = []
representation = None


def load_data(**kwargs):
    global time_values, representation
    # CLI
    args, _ = server.cli.parse_known_args()
    full_path = os.path.abspath(args.data)
    base_path = os.path.dirname(full_path)
    files = []
    reader_props = {}

    with open(full_path, "r") as f:
        metadata = json.load(f)
        reader_props = metadata.get("reader_properties", {})
        fields = metadata.get("fields", [])
        for name in metadata.get("files", []):
            files.append(os.path.join(base_path, name))

    reader = simple.OpenDataFile(files)
    for key, value in reader_props.items():
        reader.GetProperty(key).SetData(value)

    reader.UpdatePipeline()
    # Call calc_difference function
    fields, reader = calc_difference(fields, reader)
    representation = simple.Show(reader, view)
    time_values = list(time_keeper.TimestepValues)

    update_color_by(0, fields, "remote")
    state.time_value = time_values[0]
    state.times = len(time_values) - 1
    state.fields = fields

    simple.ResetCamera()
    view.CenterOfRotation = view.CameraFocalPoint
    update_view("local")


ctrl.on_server_ready.add(load_data)

# -----------------------------------------------------------------------------
# ParaView pipeline
# -----------------------------------------------------------------------------

# Rendering setup
view = simple.GetRenderView()
view.UseColorPaletteForBackground = 0
view.Background = [0.8, 0.8, 0.8]
view.OrientationAxesVisibility = 0
view = simple.Render()


def calc_difference(fields, reader):
    # Define scalar to use for the calculation
    scalar = 'accl'
    # Set the active source to the reader
    simple.SetActiveSource(reader)
    # Use the 'Force Time' filter to select t0
    forceTime = simple.ForceTime(
        registrationName='ForceTime1', Input=reader)
    # Use the 'Calculator' to name the array at t0
    calculator1 = simple.Calculator(registrationName='Calculator1',
                                    Input=forceTime)
    calculator1.AttributeType = 'Point Data'
    calculator1.ResultArrayName = 'accl_0'
    calculator1.Function = scalar
    # Use the pass arrays and append attributes filters to append the new array to the input data
    passArrays = simple.PassArrays(registrationName='PassArrays1',
                                   Input=calculator1)
    passArrays.PointDataArrays = ['accl_0']
    appendAttributes = simple.AppendAttributes(registrationName='AppendAttributes1',
                                               Input=[reader, passArrays])
    # Use another calculator to calculate the difference in the array from t0 (accl_0)
    calculator2 = simple.Calculator(registrationName='Calculator2',
                                    Input=appendAttributes)
    calculator2.AttributeType = 'Point Data'
    calculator2.ResultArrayName = 'd_accl'
    calculator2.Function = '"accl"-"accl_0"'
    # Add the new array to fields so that it can be used to color the vtks in update_color_by
    fields.append({
        "text": 'd_accl',
        "value": 3,
        "location": "POINTS",
        "range": [60000, 201941280]
    })

    reader.UpdatePipeline()

    return fields, reader


# -----------------------------------------------------------------------------
# Callbacks
# -----------------------------------------------------------------------------
# Rest of script is the same as the example...
```

Thank you!

**Comment by jourdain** [comment]
`vtkSMPVRepresentationProxy` is a server-side / ParaView class. 

So something is wrong or happening in the wrong order when executed within trame. It is hard to tell without running the full thing with data and full executable. 

But what I don't understand in your code is that you create a representation of the reader while the field you want to look at comes from a different source.

**Reply by smkacur** [reply]
@jourdain thank you for your prompt reply!

That is a good point. I created the representation of the reader because I thought the calculated array would be available there. In the calc_difference function I appendAttributes to the reader. I have also tried appending the calculated array 'd_accl' back to the reader but ended up with the same error. 

I have attached the data from the referenced [example below](https://github.com/Kitware/trame/blob/master/examples/05_charts/Plotly/01_plotly-charts-resizable.py#L222), as well as the full script, which I am running with this executable: 

'''
  /Applications/ParaView-5.11.0.app/Contents/bin/pvpython \
    /Users/skacur/trame/paraview/TimeAnimation/app_test.py \
    --venv /Users/skacur/.pyenv/versions/trame \
    --data /Users/skacur/trame/paraview/data/can.json
'''
[trame_paraview.zip](https://github.com/Kitware/trame/files/12663334/trame_paraview.zip)

**Comment by jourdain** [comment, marked as answer]
Arrays are never added to the input.. You need to show the tip of the pipeline, not the head.

**Reply by smkacur** [reply]
@jourdain this solved it, thank you!
